rustzx-core/src/emulator/snapshot/szx.rs

```diff
@@ -0,0 +1,479 @@
+    let mut block_header = [0u8; ZXST_BLOCK_HEADER_SIZE];
+    while asset.read_exact(&mut block_header).is_ok() {
+        let id: u32 = u32::from_le_bytes([
+            block_header[0],
+            block_header[1],
+            block_header[2],
+            block_header[3],
+        ]);
+        let size: u32 = u32::from_le_bytes([
+            block_header[4],
+            block_header[5],
+            block_header[6],
+            block_header[7],
+        ]);
+        let crtr_bytes = &[
+            block_header[0],
+            block_header[1],
+            block_header[2],
+            block_header[3],
+        ];
+        let id_str = from_utf8(crtr_bytes).unwrap().to_uppercase();
+        // println!("Block id: {id_str}, size: {size}");
+        cursor_pos += ZXST_BLOCK_HEADER_SIZE;
+
+        // ZXST Block Data
+        asset.seek(SeekFrom::Start(cursor_pos))?;
+        let mut block_data = vec![0; size as usize];
+
+        asset.read_exact(&mut block_data)?;
+        match id_str.as_str() {
+            "CRTR" => {
+                process_crtr_block(emulator, &block_data);
+            }
+            "Z80R" => {
+                process_z80r_block(emulator, &block_data);
+            }
+            "SPCR" => {
+                process_spcr_block(emulator, machine_id, &block_data);
+            }
+            #[cfg(all(feature = "sound", feature = "ay"))]
+            "AY\0\0" => {
+                process_ay_block(emulator, machine_id, &block_data);
+            }
+            "KEYB" => {
+                process_keyb_block(emulator, &block_data);
+            }
+            "AMXM" => {
+                process_amxm_block(emulator, &block_data);
+            }
+            "RAMP" => {
+                process_ramp_block(emulator, machine_id, &block_data)?;
+            }
+
+            _ => (),
+        }
+        // skip block data
+        cursor_pos += size as usize;
+
+        asset.seek(SeekFrom::Start(cursor_pos))?;
+    }
+    emulator.controller.refresh_memory_dependent_devices();
+    //println!("SZX file processed.");
+    Ok(())
+}
+
+/// Helper class to place emulator in the state required for
+/// snapshoting and return to normal state afterwards
+struct ScopedSnapshotState<'a, H: Host> {
+    pub emulator: &'a mut Emulator<H>,
+    pub is_48k: bool,
+}
+
+impl<'a, H: Host> ScopedSnapshotState<'a, H> {
+    fn enter(emulator: &'a mut Emulator<H>) -> Self {
+        let is_48k = emulator.settings.machine == ZXMachine::Sinclair48K;
+        if is_48k {
+            emulator.cpu.push_pc_to_stack(&mut emulator.controller);
+        }
+
+        Self { emulator, is_48k }
+    }
+}
+
+impl<'a, H: Host> Drop for ScopedSnapshotState<'a, H> {
+    fn drop(&mut self) {
+        if self.is_48k {
+            self.emulator
+                .cpu
+                .pop_pc_from_stack(&mut self.emulator.controller);
+        }
+    }
+}
+
+pub fn save<H, R>(emulator: &mut Emulator<H>, mut recorder: R) -> Result<()>
```

Comment:
This will likely produce unused variables lint.
```diff
- pub fn save<H, R>(emulator: &mut Emulator<H>, mut recorder: R) -> Result<()>
+ pub fn save<H, R>(_emulator: &mut Emulator<H>, mut _recorder: R) -> Result<()>
```

Reply:
Done